scripts/velocity.py

```diff
@@ -0,0 +1,446 @@
+
+DEFAULT_CONFIG_PATH = 'velocity.json'
+try:
+    DEFAULT_CONFIG_PATH = os.path.join(os.environ['project_root'],
+                                       DEFAULT_CONFIG_PATH)
+except KeyError:
+    log.error("Please run 'source environment' from the project root.")
+    exit(1)
+
+
+class Config(object):
+
+    def __init__(self, path: str):
+        self.path = path
+        self.config = {}
+        self.auth = self.config.setdefault('auth', {})
+        self.workspace = self.config.setdefault('workspace', {})
+
+    def load(self):
+        with open(self.path, 'r') as f:
+            self.config = json.load(f)
+            self.auth = self.config.setdefault('auth', {})
+            self.workspace = self.config.setdefault('workspace', {})
+
+    def save(self):
+        with open(self.path, 'w') as cfg:
+            json.dump(self.config, cfg, sort_keys=True, indent=4)
+        os.chmod(self.path, 0o600)
+
+    @classmethod
+    def parse_workspace_url(cls, url: str) -> Mapping[str, str]:
+        """
+        >>> workspace_url = 'https://app.zenhub.com/workspaces/foo-bar/board?repos=456,789'
+        >>> Config.parse_workspace_url(workspace_url)
+        {'name': 'foo', 'id': 'bar', 'repos': [456, 789]}
+        """
+        parsed = urllib.parse.urlparse(url)
+        workspace_slug = parsed.path.split('/')[2]
+        workspace_name, workspace_id = workspace_slug.split('-')
+        repo_ids = urllib.parse.parse_qs(parsed.query)['repos'][0].split(',')
+        return {
+            'name': workspace_name,
+            # PyGithub requires IDs as int to force repo lookup by id (and not name)
+            'id': workspace_id,
+            'repos': [int(repo_id) for repo_id in repo_ids]
+        }
+
+
+config = Config(DEFAULT_CONFIG_PATH)
+
+
+def report(workspace: str, since: datetime.datetime, until: datetime.datetime):
+    zh = ZenHubAPI(config.auth['zenhub_key'], config.auth['zenhub_key_private'])
+    gh = github.Github(config.auth['github_key'])
+    issues = []
+    prs = []
+    for repo in config.workspace[workspace]['repos']:
+        for issue in gh.get_repo(int(repo)).get_issues(state='closed', since=since):
+            if issue.closed_at > until:
+                continue
+            elif issue.pull_request:
+                prs.append(PullRequest(issue, zh))
+            else:
+                issues.append(Issue(issue, zh))
+    review_time = collections.Counter()
+    review_lines = collections.Counter()
+    story_points = collections.Counter()
+    for pr in prs:
+        story_points[pr.author] += pr.linked_issue_estimate or 0 if pr.merged else 0
+        review_time.update(pr.reviewer_time)
+        review_lines.update({r: pr.lines_changed for r in pr.reviewer_time.keys()})
+    total_points = sum(i.estimate or 0 for i in (issues + prs))
+
+    print(f'''# {workspace!r} velocity report for \
+{_human_readable_date(since, until)}
+This report covers the period between {since.strftime('%+')} and \
+{until.strftime('%+')}.
```

Comment:
```suggestion
    print(f'''# {workspace!r} velocity report for {_human_readable_date(since, until)}
This report covers the period between {since.strftime('%+')} and {until.strftime('%+')}.
```

Reply:
Suggestion retracted.